storage/innobase/row/row0ins.cc

```diff
@@ -733,6 +733,49 @@ row_ins_foreign_trx_print(
 	mysql_mutex_assert_owner(&dict_foreign_err_mutex);
 }
 
+#ifdef WITH_WSREP
+/** Produce warning on foreign key action that failed on applier.
+
+@param trx       Transaction
+@param foreign   Foreign key where error happened
+@param warn_msg  FK warning message
+@param err_msg   Error message or nullptr
+*/
+ATTRIBUTE_COLD ATTRIBUTE_NOINLINE
+void wsrep_applier_log_fk(const trx_t* trx,
+			  const dict_foreign_t* foreign,
+			  const char* warn_msg,
+			  const char* err_msg=nullptr)
+{
+  if (!wsrep_applier_log_warnings(trx->mysql_thd))
+    return;
+
+  const std::string db_table= ut_get_name(trx, foreign->foreign_table_name);
```

Comment:
We could avoid several heap memory allocation operations by invoking a lower-level function, which `ut_get_name()` is invoking in the first place. This would also save some stack space:
```c++
  /* 2 * NAME_LEN for database and table name,
  and some slack for the #mysql50# prefix and quotes */
  char buf[3 * NAME_LEN];
  const char *bufend=
    innobase_convert_name(buf, sizeof buf, foreign->foreign_table_name,
                          strlen(foreign->foreign_table_name), trx->mysql_thd);
  WSREP_WARN("table %.*s", int(bufend - buf), buf);
```